src/Engine/Component/ParticleComponents/ParticleEffect.cpp

```diff
@@ -0,0 +1,393 @@
+            p->life += dt;
+
+        }
+    }
+    else if (m_effectParams->loop) {
+        if (m_effectParams->rate == 0) {
+            m_life = 0.0f;
+            m_particles = generateParticles();
+            m_activeParticlePositions = getActiveParticlePositions();
+            m_activeParticleOrientationIDs = getActiveParticleOrientationIDs();
+            m_activeMap = getActiveMap();
+            m_nextActivation = getNextActivation();
+            m_nextAttenuation = 0.0f;
+        }
+        else {
+            m_effectParams->effectDuration += m_life;
+        }
+        update(dt);
+    }
+}
+
+
+
+void ParticleEffect::updateActiveParticles(float dt) {
+    
+    if (m_effectParams->rate != 0.0f) {
+        while ((int)m_activeParticlePositions->size() > 0) {
+            Particle *p = m_particles[m_activeMap[0]];
+            if (p->life > m_effectParams->particleDuration) {
+                removeActiveParticle(0);
+            }
+            else {
+                break;
+            }
+        }
+        // Activate new particles based on rate. Cap at n limit and Remove the oldest Particle
+        while (m_life> m_nextActivation) {
+            if (m_activeParticlePositions->size() == m_effectParams->n) {
+                int pid = m_activeMap[0];
+                removeActiveParticle(0);
+                m_particles[pid] = makeParticle(pid);
+                addActiveParticle(pid);
+            }
+            else {
+                int pid;
+                if (m_activeMap.size() == 0) {
+                    pid = 0;
+                }
+                else {
+                    pid = m_activeMap.back() + 1;
+                    if (pid == m_effectParams->n) {
+                        pid = 0;
+                    }
+                }
+                m_particles[pid] = makeParticle(pid);
+                addActiveParticle(pid);
+            }
+            m_nextActivation += (1.0f / m_effectParams->rate);
+        }
+
+    }
+}
+
+void ParticleEffect::updatePosition(Particle *p, float dt, float attenuationFactor) {
+    p->velocity *= attenuationFactor;
+    for (int i = 0; i < (int)m_effectParams->accelerators->size(); i++) {
+        p->velocity += m_effectParams->accelerators->at(i) * dt;
+    }
+    p->position += p->velocity * dt; 
+}
+
+int ParticleEffect::getAttenuations(float dt) {
+    if (m_effectParams->attenuation == 1.0f) {
+        return 0;
+    }
+    else {
+        int attenuations = 0;
+        while (m_life > m_nextAttenuation) {
+            attenuations += 1;
+            m_nextAttenuation += m_attenuationRate;
+        }
+        return attenuations;
+    }
+}
+
+ //Remove from specific location
+void ParticleEffect::removeActiveParticle(int i) {
+    m_activeMap.erase(m_activeMap.begin() + i);
+    m_activeParticlePositions->erase(m_activeParticlePositions->begin() + i);
+    m_activeParticleOrientationIDs->erase(m_activeParticleOrientationIDs->begin() + i);
+}
```

Comment:
Do we need to call delete here?

Reply:
No, there are a list of active particle positions to be passed to the shader, and this removes those positions from that list. Deletion should only happen at the end of a particle effect if ever.